From 52df783274946146922ac5002fab121928d0344f Mon Sep 17 00:00:00 2001
From: Dmitry Kasatkin <d.kasatkin@samsung.com>
Date: Wed, 7 May 2014 16:22:40 +0300
Subject: [PATCH 39/45] ima: return dentry name if mnt is NULL

This patch returns dentry name if mnt is NULL.
Symbolic links checking hook does not have mnt.

Signed-off-by: Dmitry Kasatkin <d.kasatkin@samsung.com>
---
 security/integrity/ima/ima_api.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/security/integrity/ima/ima_api.c b/security/integrity/ima/ima_api.c
index c81128b..29b2d82 100644
--- a/security/integrity/ima/ima_api.c
+++ b/security/integrity/ima/ima_api.c
@@ -315,9 +315,12 @@ void ima_audit_measurement(struct integrity_iint_cache *iint,
 const char *ima_d_path(struct path *path, char **pathbuf)
 {
 	char *pathname = NULL;
+	struct dentry *dentry = path->dentry;
 
-	if (!path->mnt)
+	if (!dentry)
 		return NULL;
+	if (!path->mnt)
+		goto out;
 
 	*pathbuf = __getname();
 	if (*pathbuf) {
@@ -328,7 +331,8 @@ const char *ima_d_path(struct path *path, char **pathbuf)
 			pathname = NULL;
 		}
 	}
-	return pathname ?: (const char *)path->dentry->d_name.name;
+out:
+	return pathname ?: (const char *)dentry->d_name.name;
 }
 
 static int prepend(char **buffer, int buflen, const char *str, int namelen)
-- 
2.1.4

